src/Symfony/Component/Messenger/MessageBus.php

```diff
@@ -35,17 +35,27 @@ public function __construct(iterable $middlewareHandlers = [])
         } elseif (\is_array($middlewareHandlers)) {
             $this->middlewareAggregate = new \ArrayObject($middlewareHandlers);
         } else {
-            $this->middlewareAggregate = new class() {
-                public $aggregate;
-                public $iterator;
+            // This class expects 'middlewareAggregate' property is an instance
+            // of IteratorAggregate So we wrap $middlewareHandlers in to it, and
+            // we cache the results in order to work even with generator.
```

Comment:
Update proposal:
$this->middlewareAggregate should be an instance of IteratorAggregate.
When $middlewareHandlers is an Iterator, we wrap it to ensure it is lazy-loaded and can be rewound.

Reply:
Perfect :ok_hand: I updated the PR